File: app/src/main/java/ru/cool/lwjgl/controls/CameraController.java
package ru.cool.lwjgl.controls;

import org.joml.Matrix4f;
import org.joml.Vector3f;
import ru.cool.lwjgl.Config;
import ru.cool.lwjgl.input.Buttons;
import ru.cool.lwjgl.input.Keyboard;
import ru.cool.lwjgl.input.Mouse;
import ru.cool.lwjgl.objects.Camera;
import ru.cool.lwjgl.utils.Time;

public class CameraController extends MovingObjectsController{

    private Camera camera;

    public CameraController(Camera camera) {
        super(camera);
        this.camera = camera;
    }
    private void moveCamera(){
        float cameraSpeed = (float) (Time.deltaTime * camera.getSpeed());
        if (Keyboard.isButtonPress(Buttons.BACK)){
            camera.getPosition().add(new Vector3f(0,0, cameraSpeed).sub(camera.getDirection()).normalize().mul(cameraSpeed));
        }
        if (Keyboard.isButtonPress(Buttons.FORWARD)){
            camera.getPosition().add(new Vector3f(0,0, cameraSpeed).add(camera.getDirection()).normalize().mul(cameraSpeed));
        }
        if (Keyboard.isButtonPress(Buttons.RIGHT)){
            camera.getPosition().add(new Vector3f(camera.getDirection()).cross(camera.getUpVector()).normalize().mul(cameraSpeed));
        }
        if (Keyboard.isButtonPress(Buttons.LEFT)){
            camera.getPosition().sub(new Vector3f(camera.getDirection()).cross(camera.getUpVector()).normalize().mul(cameraSpeed));
        }
        if (Keyboard.isButtonPress(Buttons.UP)){
            camera.getPosition().add(new Vector3f(0, cameraSpeed,0));
        }
        if (Keyboard.isButtonPress(Buttons.DOWN)){
            camera.getPosition().sub(new Vector3f(0, cameraSpeed,0));
        }
    }

    private void rotateCamera(){
        float xOffset = (Mouse.getMousePosX() - Mouse.lastX);
        float yOffset = (Mouse.getMousePosY() - Mouse.lastY);
        Mouse.lastX = (int) Mouse.getMousePosX();
        Mouse.lastY = (int) Mouse.getMousePosY();

        xOffset *= Config.SENSITIVITY;
        yOffset *= Config.SENSITIVITY;

        camera.setRotationYaw(camera.getRotationYaw() + xOffset);
        camera.setRotationPitch((float) Math.clamp(camera.getRotationPitch() + yOffset, -89.9, 89.9));

        Vector3f direction = new Vector3f();
        direction.x = (float) ((float) Math.cos(Math.toRadians(camera.getRotationPitch())) * Math.cos(Math.toRadians(camera.getRotationYaw())));
        direction.y = -(float) Math.sin(Math.toRadians(camera.getRotationPitch()));
        direction.z = (float) ((float) Math.cos(Math.toRadians(camera.getRotationPitch())) * Math.sin(Math.toRadians(camera.getRotationYaw())));

        camera.setDirection(direction.normalize());
        camera.setViewMatrix(new Matrix4f().lookAt(camera.getPosition(), new Vector3f(camera.getPosition()).add(camera.getDirection()), camera.getUpVector()));
    }

    public Camera getCamera() {
        return camera;
    }

    public void setCamera(Camera camera) {
        this.camera = camera;
    }

    @Override
    public void invoke() {
        this.moveCamera();
        this.rotateCamera();
    }

}

File: kotlin_app/src/main/kotlin/ru/cool/lwjgl_kotlin/input/controllers/CameraController.kt
package ru.cool.lwjgl_kotlin.input.controllers

import ru.cool.lwjgl_kotlin.input.Buttons
import ru.cool.lwjgl_kotlin.input.Keyboard
import ru.cool.lwjgl_kotlin.input.Mouse
import ru.cool.lwjgl_kotlin.objects.camera.PerspectiveCamera
import ru.cool.lwjgl_kotlin.utils.Time
import ru.cool.lwjgl_kotlin.Config
import kotlin.math.cos
import kotlin.math.sin
import org.joml.Math
import org.joml.Vector3f
import ru.cool.lwjgl_kotlin.*

class CameraController(var camera: PerspectiveCamera, var speed: Float): IController {
    override fun invoke() {
        moveCamera()
        rotateCamera()
    }

    fun rotateCamera(){
        val mouseX = (Mouse.getMousePosX() - Mouse.lastX) * Config.SENSITIVITY
        val mouseY = (Mouse.getMousePosY() - Mouse.lastY) * Config.SENSITIVITY

        Mouse.lastX = Mouse.getMousePosX().toInt()
        Mouse.lastY = Mouse.getMousePosY().toInt()

        camera.cameraRotationYaw += mouseX
        camera.cameraRotationPitch += mouseY

        camera.cameraRotationPitch = Math.clamp(-89f, 89f, camera.cameraRotationPitch)

        val directionX = (Math.cos(Math.toRadians(camera.cameraRotationPitch)) * Math.cos(Math.toRadians(camera.cameraRotationYaw)));
        val directionY = -Math.sin(Math.toRadians(camera.cameraRotationPitch));
        val directionZ = (Math.cos(Math.toRadians(camera.cameraRotationPitch)) * Math.sin(Math.toRadians(camera.cameraRotationYaw)));

        val direction = Vector3f(directionX, directionY, directionZ)
        camera.direction = direction
        camera.updateMatrix()
    }

    fun moveCamera() {
        val cameraSpeed = speed * Time.deltaTime.toFloat()
        if (Keyboard.isButtonPress(Buttons.W)) {
            camera.move(Vector3f(0f, 0f, cameraSpeed).sub(camera.direction).normalize().mul(cameraSpeed))
        }
        if (Keyboard.isButtonPress(Buttons.S)) {
            camera.move(Vector3f(0f, 0f, cameraSpeed).add(camera.direction).normalize().mul(cameraSpeed))
        }
        if (Keyboard.isButtonPress(Buttons.A)){
            camera.move(Vector3f(camera.direction).cross(camera.upVector) * cameraSpeed)
        }
        if (Keyboard.isButtonPress(Buttons.D)){
            camera.move(Vector3f(camera.direction).cross(camera.upVector) * -cameraSpeed)
        }
        if (Keyboard.isButtonPress(Buttons.LEFT_CTR)){
            camera.move(0f,cameraSpeed,0f)
        }
        if (Keyboard.isButtonPress(Buttons.SPACE)){
            camera.move(0f,-cameraSpeed,0f)
        }
        if (Keyboard.isButtonPress(Buttons.R)){
            camera.setPos(Vector3f(0f,0f,3f))
        }
    }
}